### DISCUSSION TITLE
Consider if the required and optional columns are right

### DISCUSSION THREAD

**Discussion by francisbarton** [ThomUK/SPCreporter#77] [General]
I wonder if we can do one of the following two, opposite strategies:

a) Be really opinionated on which columns need to be included in the measure config file, and just make all potential columns compulsory
b) Only make essential columns (ref, measure_name, unit, improvement_direction?) required, and make everything else optional. It strikes me that things like target and owner are good to have but strictly speaking optional, in that the report and charts can be successfully generated without them. And then the package will just add on any missing optional cols (as NAs), and we then just have to ensure that NA values are handled appropriately by all the package processes.

**Comment by ThomUK** [comment]
Thanks for this, and for the first package "github discussion".

a) Is nice and simple for package development, but is a breaking change for users of existing spreadsheets whenever new functionality that depends on a new config column is released.  
b) Is more involved, but makes for a better user experience.  There would be a "minimalist" use of the package, with additional functionality bundled in as standard that can be ignored if not used.  

I reluctantly lean towards implementing b.  It is more work, but the main use-case of this package is customer reporting by users who may not be "power users" of R.  Breaking changes as we develop will cause error messages during the course of reporting use, which is unsettling at best, and a showstopper for a critical report deadline at worst.  

I think the most secure way to implement b would be to drive the behaviour in with tests.  ie. proper TDD where we define the result we need at high level (no errors with a minimal column set passed in), and also specify the detail function by function with tests at unit level.  I'd like to make sure the package errors hard and early for unexpected input, together with helpful error messages specifying in as much detail as necessary how to fix the problem (aimed at new R users).  I think we should avoid coercion between character and numeric types.  Even correcting capitalisation silently is something we need to think carefully about, to avoid feeling too "magic".
